pkg/storage/unified/sql/notifier_sql.go

```diff
@@ -0,0 +1,222 @@
+	errInvalidWatchBufferSize = fmt.Errorf("watchBufferSize must be greater than 0")
+	errInvalidPollingInterval = fmt.Errorf("pollingInterval must be greater than 0")
+	errDoneRequired           = fmt.Errorf("done is required")
+	errDialectRequired        = fmt.Errorf("dialect is required")
+)
+
+// pollingNotifier is a notifier that polls the database for new events.
+type pollingNotifier struct {
+	dialect         sqltemplate.Dialect
+	pollingInterval time.Duration
+	watchBufferSize int
+
+	log    log.Logger
+	tracer trace.Tracer
+
+	batchLock     *batchLock
+	listLatestRVs func(ctx context.Context) (groupResourceRV, error)
+	historyPoll   func(ctx context.Context, grp string, res string, since int64) ([]*historyPollResponse, error)
+
+	done <-chan struct{}
+}
+
+type pollingNotifierConfig struct {
+	dialect         sqltemplate.Dialect
+	pollingInterval time.Duration
+	watchBufferSize int
+
+	log    log.Logger
+	tracer trace.Tracer
+
+	batchLock     *batchLock
+	listLatestRVs func(ctx context.Context) (groupResourceRV, error)
+	historyPoll   func(ctx context.Context, grp string, res string, since int64) ([]*historyPollResponse, error)
+
+	done <-chan struct{}
+}
+
+func (cfg *pollingNotifierConfig) validate() error {
+	if cfg.historyPoll == nil {
+		return errHistoryPollRequired
+	}
+	if cfg.listLatestRVs == nil {
+		return errListLatestRVsRequired
+	}
+	if cfg.batchLock == nil {
+		return errBatchLockRequired
+	}
+	if cfg.tracer == nil {
+		return errTracerRequired
+	}
+	if cfg.log == nil {
+		return errLogRequired
+	}
+	if cfg.watchBufferSize <= 0 {
+		return errInvalidWatchBufferSize
+	}
+	if cfg.pollingInterval <= 0 {
+		return errInvalidPollingInterval
+	}
+	if cfg.done == nil {
+		return errDoneRequired
+	}
+	if cfg.dialect == nil {
+		return errDialectRequired
+	}
+	return nil
+}
+
+func newPollingNotifier(cfg *pollingNotifierConfig) (*pollingNotifier, error) {
+	if err := cfg.validate(); err != nil {
+		return nil, fmt.Errorf("invalid polling notifier config: %w", err)
+	}
+	return &pollingNotifier{
+		dialect:         cfg.dialect,
+		pollingInterval: cfg.pollingInterval,
+		watchBufferSize: cfg.watchBufferSize,
+		log:             cfg.log,
+		tracer:          cfg.tracer,
+		batchLock:       cfg.batchLock,
+		listLatestRVs:   cfg.listLatestRVs,
+		historyPoll:     cfg.historyPoll,
+		done:            cfg.done,
+	}, nil
+}
+
+func (p *pollingNotifier) notify(ctx context.Context) (<-chan *resource.WrittenEvent, error) {
+	since, err := p.listLatestRVs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("watch, get latest resource version: %w", err)
+	}
+	stream := make(chan *resource.WrittenEvent, p.watchBufferSize)
+	go p.poller(ctx, since, stream)
+	return stream, nil
+}
+
+func (p *pollingNotifier) poller(ctx context.Context, since groupResourceRV, stream chan<- *resource.WrittenEvent) {
+	t := time.NewTicker(p.pollingInterval)
+	defer close(stream)
+	defer t.Stop()
+	isSQLite := p.dialect.DialectName() == sqltemplate.SQLite.DialectName()
```

Comment:
we can skip this if we never enable the pollingNotifier for SQLite

Reply:
Changed: 3adab8d9b6a0daa95a74ca04d4eaa6e9acbcaabd